File: py/classic/greedy_algorithm/KnapsackOfFractional.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import functools
from typing import List

class KnapsackOfFractional:
    def solution(self, weights: List[int], values: List[int], target: int) -> int:
        goods = []
        for weight, value in zip(weights, values):
            goods.append(self.__GoodNode(weight, value))

        goods.sort(key = functools.cmp_to_key(self.__cmp))

        result = 0
        for good in goods:
            if target < good.weight:
                return result + target * good.value // good.weight

            target -= good.weight
            result += good.value
        return result

    def __cmp(self, node1: '__GoodNode', node2: '__GoodNode') -> int:
        return 1 if node1.rate < node2.rate else -1

    class __GoodNode:
        def __init__(self, weight: int, value: int):
            self.weight, self.value, self.rate = weight, value, value / weight

File: cpp/classic/greedy_algorithm/KnapsackOfFractional.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class KnapsackOfFractional {
public:
    int solution(vector<int> &weights, vector<int> &values, int target) {
        vector<GoodNode *> goods(0);
        for (int i = 0; i < (int) weights.size(); ++i) {
            goods.emplace_back(new GoodNode(weights[i], values[i]));
        }

        sort(goods.begin(), goods.end(), cmp);

        int result = 0;
        for (GoodNode *good : goods) {
            if (target < good->weight) {
                return result + target * good->value / good->weight;
            }

            target -= good->weight;
            result += good->value;
        }
        return result;
    }

private:
    typedef struct GoodNode {
        int weight;
        int value;
        double rate;

        GoodNode(int weight, int value) : weight(weight), value(value), rate((double) value / weight) {}
    } GoodNode, *pNode;

    static bool cmp(const pNode &node1, const pNode &node2) {
        return node1->rate > node2->rate;
    }
};
